src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/SyntaxNodeExtensions.cs

```diff
@@ -189,27 +192,101 @@ public static int FullWidth(this SyntaxNode node)
+            => nodes.FindInnermostCommonNode()?.FirstAncestorOrSelf(predicate);
+
+        public static SyntaxNode? FindInnermostCommonNode(this IEnumerable<SyntaxNode> nodes)
         {
-            IEnumerable<SyntaxNode>? blocks = null;
+            // Two collections we use to make this operation as efficient as possible. One is a
+            // stack of the current shared ancestor chain shared by all nodes so far.  It starts
+            // with the full ancestor chain of the first node, and can only get smaller over time.
+            // It should be log(n) with the size of the tree as it's only storing a parent chain.
+            //
+            // The second is a set with the exact same contents as the array.  It's used for O(1)
+            // lookups if a node is in the ancestor chain or not.
+
+            using var _1 = ArrayBuilder<SyntaxNode>.GetInstance(out var commonAncestorsStack);
+            using var _2 = PooledHashSet<SyntaxNode>.GetInstance(out var commonAncestorsSet);
+
+            var first = true;
             foreach (var node in nodes)
             {
-                blocks = blocks == null
-                    ? node.AncestorsAndSelf().Where(predicate)
-                    : blocks.Intersect(node.AncestorsAndSelf().Where(predicate));
+                // If we're just starting, initialize the ancestors set/array with the ancestors of
+                // this node.
+                if (first)
+                {
+                    first = false;
+                    foreach (var ancestor in node.ValueAncestorsAndSelf())
+                    {
+                        commonAncestorsSet.Add(ancestor);
+                        commonAncestorsStack.Add(ancestor);
+                    }
+
+                    // Reverse the ancestors stack so that we go downwards with CompilationUnit at
+                    // the start, and then go down to this starting node.  This enables cheap
+                    // popping later on.
+                    commonAncestorsStack.ReverseContents();
+                    continue;
+                }
+
+                // On a subsequent node, walk its ancestors to find the first match
+                var commonAncestor = FindCommonAncestor(node, commonAncestorsSet);
+                if (commonAncestor == null)
+                {
+                    // So this shouldn't happen as long as the nodes are from the same tree.  And
+                    // the caller really shouldn't be calling from different trees.  However, the
+                    // previous impl supported that, so continue to have this behavior.
+                    //
+                    // If this doesn't fire, that means that all callers seem sane.  If it does
+                    // fire, we can relax this (but we should consider fixing the caller).
+                    Debug.Fail("Could not find common ancestor.");
```

Comment:
note: this assert is mostly for testing.  if tests pass, i want to change this to just throw and not return null.